sdk/src/main/java/co/omise/android/api/Invocation.kt

```diff
@@ -0,0 +1,78 @@
+package co.omise.android.api
+
+import android.os.Handler
+import co.omise.android.models.APIError
+import co.omise.android.models.Model
+import co.omise.android.models.ModelParserUtil.parseModelFromJson
+import okhttp3.Call
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import org.json.JSONException
+import java.io.IOException
+
+internal class Invocation<T : Model>(
+        private val replyHandler: Handler,
+        private val httpClient: OkHttpClient,
+        private val request: co.omise.android.api.Request<T>,
+        private val listener: RequestListener<T>
+) {
+
+    fun invoke() {
+        val call = httpClient.newTypedCall(
+                Request.Builder()
+                        .method(request.method,
+                                request.payload)
+                        .url(request.url)
+                        .build(),
+                request.responseType)
+
+        try {
+            processCall(call)
+        } catch (e: IOException) {
+            didFail(e)
+        } catch (e: JSONException) {
+            didFail(e)
+        }
+
+    }
+
+    private fun processCall(call: TypedCall) {
+        val response = call.execute()
+        if (response.body() == null) {
+            didFail(IOException("HTTP response have no body."))
+            return
+        }
+
+        val rawJson = response.body().string()
+        val model = parseModelFromJson(rawJson, call)
+        if (response.code() in 200..299 && model != null) {
```

Comment:
I think we should have a special case for handling 300..399 status codes? We shouldn't allow redirection and have a specific case.

Reply:
So, a custom exception?

Reply:
Correct

Reply:
What would we call this exception btw? :D